### PR TITLE
Make builder pattern more practical

### PR DESCRIPTION
**Pull request by TGotwig**
https://github.com/TGotwig/patterns/blob/patch-1/patterns/builder.md

### CODE REVIEW

**Comment by nesteruk** [comment]
It is confusing to have `bar: String` in both `Foo` and `FooBuilder` — consider renaming one of the `bar`s to something else.

**Comment by TGotwig** [comment]
@nesteruk Changed this and a little bit more 🙂

File: patterns/builder.md
Comment (lambda-fairy) [inline patterns/builder.md:23]: This `Display` impl feels irrelevant to illustrating the builder pattern. Can we replace this with `#[derive(Debug)]`, and use `format!("{:?}", foo)` in the test instead?

**Review by lambda-fairy — DISMISSED** [review]
LGTM, modulo the comment below

File: patterns/builder.md
Comment (TGotwig) [reply patterns/builder.md:23]: I will look for this soon 🙂

**Comment by pickfire** [comment]
This example does not seemed to be explaining much, maybe we could have a real-life example or an explanation on why it needs to be constructed this way rather than just specifying the struct directly?

File: patterns/builder.md
Comment (TGotwig) [reply patterns/builder.md:14]: Yeah you are right! Changed this up 👷

**Comment by TGotwig** [comment]
@pickfire Hmm, changed it a little bit up.
I think that runnable code is better than not runnable code and the comments should explain enough 🧐

File: patterns/builder.md
Comment (pickfire) [inline patterns/builder.md:37]: Isn't it usually `build`?

**Review by pickfire — COMMENTED** [review]
The indentation seemed off, should be 4 spaces rather than 2 spaces.

File: patterns/builder.md
Comment (pickfire) [inline patterns/builder.md:18]: We can probably just keep the same `bar` here.

File: patterns/builder.md
Comment (pickfire) [inline patterns/builder.md:16]: Why shouldn't we keep the builder name as `FooBuilder`?

File: patterns/builder.md
Comment (pickfire) [inline patterns/builder.md:37]: Some parts seemed inconsistent, some uses `finish` and below uses `build`.

Comment (TGotwig) [reply patterns/builder.md]: Unfortunately I had this overwritten, but now it is `build` again 👷

File: patterns/builder.md
Comment (TGotwig) [reply patterns/builder.md]: On second thought you are right, changed it back ↩️

File: patterns/builder.md
Comment (TGotwig) [reply patterns/builder.md]: Changed this up 👷

File: patterns/builder.md
Comment (pickfire) [inline patterns/builder.md:20]: Why not just do derive `PartialEq` above? Rather than writing this from scratch?

File: patterns/builder.md
Comment (pickfire) [inline patterns/builder.md:54]: Do we need this type signature, isn't it inferred?

File: patterns/builder.md
Comment (pickfire) [inline patterns/builder.md:37]: Is this a good pattern to clone within the function when users can provide `String`? I am not quite sure how this is done in the wild.

File: patterns/builder.md
Comment (TGotwig) [reply patterns/builder.md:48]: It's not needed but best practice to do so, I think 🤔

Comment (pickfire) [reply patterns/builder.md]: I am not sure if it is best practice but for docs examples (including rust), most types are explicitly state to let others to understand easily but I think this example should be way to simple, if they get to this page they should already be able to see through this simple example.

**Review by pickfire — APPROVED** [review]
Looks good to me. I think it would be good to have one of the maintainer to check this though.

**Comment by simonsan** [comment]
@TGotwig Hey, thank you for the PR. Could you please rebase on `master` so the CI is checking this through while we get to reviewing this PR? Thank you ;-)

**Comment by TGotwig** [comment]
@simonsan Sure, will do! 😃

**Comment by simonsan** [comment]
> @simonsan Sure, will do! 😃

Sorry, haven't seen that there is no addition to summary or anything needed. So I could do it myself. ;-) It's fine then.

**Comment by simonsan** [comment]
@MarcoIeni You also want to look over again?

**Review by MarcoIeni — APPROVED** [review]

**Comment by simonsan** [comment]
Thanks! :-)

**Comment by TGotwig** [comment]
Cool! 😊🚀

**Pull request merged**
